datafusion/expr/src/type_coercion/functions.rs

```diff
@@ -744,6 +706,57 @@ fn get_valid_types(
     Ok(valid_types)
 }
 
+/// Validates that all data types are either [`NativeType::String`] or [`NativeType::Null`].
+/// For [`NativeType::Null`], returns [`DataType::Utf8`] as the default string type.
+/// Returns error if any type is neither [`NativeType::String`] nor [`NativeType::Null`].
+fn validate_and_collect_string_types(data_types: &[DataType]) -> Result<Vec<DataType>> {
+    data_types
+        .iter()
+        .map(|data_type| {
+            let logical_type: NativeType = data_type.into();
+            match logical_type {
+                NativeType::String => Ok(data_type.to_owned()),
+                // TODO: Switch to Utf8View if all the string functions supports Utf8View
+                NativeType::Null => Ok(DataType::Utf8),
+                _ => plan_err!("The signature expected NativeType::String but received {logical_type}"),
+            }
+        })
+        .collect()
+}
+
+/// Returns a common string [`DataType`] that both input types can be coerced to.
+/// Handles [`DataType::Dictionary`] by recursively finding common type of their value [`DataType`].
+/// Returns error if types cannot be coerced to a common string [`DataType`].
+fn find_common_string_type(lhs_type: &DataType, rhs_type: &DataType) -> Result<DataType> {
+    match (lhs_type, rhs_type) {
+        (DataType::Dictionary(_, lhs), DataType::Dictionary(_, rhs)) => {
+            find_common_string_type(lhs, rhs)
+        }
+        (DataType::Dictionary(_, v), other) | (other, DataType::Dictionary(_, v)) => {
+            find_common_string_type(v, other)
+        }
+        _ => {
+            if let Some(coerced_type) = string_coercion(lhs_type, rhs_type) {
+                Ok(coerced_type)
+            } else {
+                plan_err!(
+                    "{lhs_type} and {rhs_type} are not coercible to a common string type"
+                )
+            }
+        }
+    }
+}
+
+/// Recursively traverses [`DataType::Dictionary`] to get the underlying value [`DataType`].
+/// For non-dictionary types, returns the default [`DataType`].
+fn base_dictionary_type_or_default_type(data_type: &DataType) -> DataType {
```

Comment:
We can add dictionary support to `datafusion/common/src/utils/mod.rs` `base_type`